src/index.rs

```diff
@@ -1,132 +1,278 @@
+            'token_loop: for (token, ids) in vocabulary.tokens_to_ids().iter() {
+                if ids.contains(&eos_token_id) {
+                    continue;
+                }
+
+                let mut next_state = current_state;
+                for transition_byte in token {
+                    next_state = dfa.next_state(next_state, *transition_byte);
+                    if dfa.is_dead_state(next_state) || dfa.is_quit_state(next_state) {
+                        continue 'token_loop;
+                    }
+                }
+
+                let is_intermediate_state = !dfa.is_match_state(next_state);
+                let is_full_match_state = dfa.is_match_state(dfa.next_eoi_state(next_state));
+                if is_intermediate_state || is_full_match_state {
+                    for token_id in ids {
+                        transitions
+                            .entry(current_state.as_u32())
+                            .or_default()
+                            .insert(*token_id, next_state.as_u32());
+                    }
+                }
+                if !seen.contains(&next_state) {
+                    seen.insert(next_state);
+                    next_states.push(next_state);
+                }
             }
+        }
 
-            seen.insert(start_state);
+        // Populate `transitions` with mappings from `final_states` to `eos_token_id`
+        for &final_state in &final_states {
+            transitions
+                .entry(final_state)
+                .or_default()
+                .insert(eos_token_id, final_state);
         }
 
-        let is_valid = states_to_token_subsets
-            .values()
-            .flat_map(|token_id_end_states| token_id_end_states.values())
-            .any(|end_state| fsm_info.finals.contains(end_state));
+        // Check if there is at least one valid mapping
+        let is_valid = transitions.values().any(|mapping| {
+            mapping
+                .values()
+                .any(|end_state| final_states.contains(end_state))
+        });
 
         if is_valid {
             Ok(Self {
-                initial: fsm_info.initial,
-                finals: fsm_info.finals.clone(),
-                states_to_token_subsets,
+                initial_state: start_state.as_u32(),
+                final_states,
+                transitions,
                 eos_token_id,
             })
         } else {
-            Err(Error::IndexError)
+            Err(Error::InsufficientVocabulary)
         }
     }
 
-    pub(crate) fn allowed_tokens(&self, state: u32) -> Option<Vec<u32>> {
-        self.states_to_token_subsets
-            .get(&state)
+    /// Lists allowed tokens for a give state ID or `None` if it is not found in `Index`.
+    pub fn allowed_tokens(&self, state: &StateId) -> Option<Vec<TokenId>> {
+        self.transitions
+            .get(state)
             .map_or_else(|| None, |res| Some(res.keys().cloned().collect()))
```

Comment:
`.map_or_else(|| None, |_| Some(...))` is `.map(|_| ...)`

Could you try `.map` here?

Reply:
I think so, we're only dealing with the `Some` case and don't need a special fallback for the `None` case, it will simplify reading 👍